.github/workflows/pr-build.yml

```diff
@@ -61,8 +63,42 @@ jobs:
           flags: integration
 
       - name: push docker image
-        if: github.event.pull_request.merged
+         # https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#pull_request
+        if: github.event_name == 'pull_request' && github.event.action == 'closed' && github.event.pull_request.merged == true
         run: ./gradlew dockerPushImages
 
+  validate-helm-charts:
```

Comment:
@kotharironak - I disagree with this consolidation.

Keeping the jobs separate has a few different benefits
- They run in parallel, giving a faster time to green build
- The reported results are more granular, making it easier for the developer to figure out what went wrong
- The jobs are kept independent and don't unintentionally use state from a previous step
- Developers get more feedback because a serial run would stop at the first failure, while a parallel run gives the results for each

The trade offs that I see are only:
- Takes more action minutes (not an issue at the moment, this is a free project)
- More maintenance (I actually think keeping workflows simple is actually easier to maintain even if more files/LOC)

Reply:
As per github actions docs,
- jobs run in parallel if the workflow has multiple jobs- https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#jobs
- in Github hosted runner each job run in fresh virtual envs - https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#runners
So, I think combing multiple jobs that trigger on the same condition will be easy to manage directly from the action tab. With the above changes - https://github.com/hypertrace/attribute-service/actions/runs/430169794 (Total time ~2m)

Below was the previous case where we had three workflows on the same triggers and three things to look at (Total time ~3min)
pr-build.yml - https://github.com/hypertrace/attribute-service/actions/runs/430001233
synk.yml - https://github.com/hypertrace/attribute-service/actions/runs/430001235
validate-charts.yml - https://github.com/hypertrace/attribute-service/actions/runs/430001234

Reply:
Sorry I misunderstood the ask - I thought we were combining jobs, not workflows. Combining the workflows based on the triggers makes sense, so disregard. Really what I was looking for is that they show up independently on the PR itself, which they appear to be in this change. I think optimizing those screens - that is 
```
https://github.com/hypertrace/attribute-service/pull/45
https://github.com/hypertrace/attribute-service/pull/45/checks
```
is the much more common use case vs looking at the repo-wide action runs (but sounds like we've got both!)

Reply:
Given this though, does it make sense to rename the workflows based on those triggers rather than their content?

Reply:
Yes, that is what I was suggesting @JBAhire  as well. I was also suggesting that instead of running both the targets on each pull request (`pull_request_target` and `pull_request`) we can have three workflows.

1. pr-build-forked.yml (runs only for forked branches) - in `pull_request_target` context
2. pr-build.yml (runs only for member's pull request) - in `pull_request` context
3. publish.yml (only for member pull request) - which is anyway the case as it triggers on `release`

Most of the steps will be the same (and may look the same copy) between the `pr-build-forked` and `pr-build` workflow, but, we will able control what goes in forked PRs workflow. And, makes it easy to separate issues (like secrets or anything).

Reply:
Doesn't `pull_request_target` run on branch + fork PRs? (for example, it ran on this PR which is a branch). I was thinking of getting rid of the `pull_request` trigger entirely, and instead having
- `pull_request_target` - everything we expect on a PR, which should be the same for members or contributors
- `push` - everything we expect to run on merge (basically, the "merge-publish" from before - only triggered by members who merge)
- `release` - everything for cutting a release (only members who cut a release would be able to trigger)

Reply:
so there are 2 problems with using `pull_request_target` for this workflow.
- codecov action error which you can see here as well. (https://github.com/codecov/codecov-action/issues/155)
- checkout like this: https://github.com/hypertrace/attribute-service/pull/45#discussion_r545827000

Reply:
ideally only `pull_request_target` would be enough for most of the situations and that's what many projects follow, trigger on `push` and `pull_request_target`.

Reply:
- checkout: This is just a change in the checkout command right? Does it have other implications or is that resolved once we make that change?

- codecov: If `pull_request_target` doesn't work for upload, would it make sense to create a fourth workflow
`pull_request` with one job, `test` that runs the tests and uploads coverage. That shouldn't need secrets, right?

Reply:
Basically, I'm interested if we can avoid duplicating between the `pull_request_target` and `pull_request` workflows, and just sort the jobs based on what they need in terms of available info.

Reply:
The reason I was suggesting above (forked vs non-forked flow)
a. avoiding running both (pull_request, pull_request_target) on the same PR (but your proposal can also handle it)
b. educating folks (members) on `pull_request_target` and `pull_request` for security, and if some repo's are having fewer contributions, we are still running most of the PRs in `pull_request_target`
c. If we don't want to run certain steps in `forked` PR like Snyk (this was under discussion), it can help.
d. resource utilization (`test` workflow anyway depends on `build`) and better workflow view (https://github.com/hypertrace/attribute-service/actions/runs/430169794)

But, I am fine with your suggestion too. That will also help in codcov issue for now which we might have to omit in (pr-build-forked, pr-build) proposal.

Just to summerize:
pr-build - on `pull_reqeust_target` + `push`
pr-test - on `pull_reqeuest` + `push`
merge_publish - on push
publish - on release tag created

Reply:
will create this 4 workflows.